From 13aac73e687b6bfc25ca68051955866d94b59e2e Mon Sep 17 00:00:00 2001
From: Ferry Toth <ftoth@exalondelft.nl>
Date: Fri, 19 Jun 2026 22:28:32 +0200
Subject: [PATCH 1/1] Add shared make jobserver support

This is a rework of Richard Purdie's patch at
http://git.yoctoproject.org/cgit.cgi/poky-contrib/commit/?h=rpurdie/wipqueue4&id=d66a327fb6189db5de8bc489859235dcba306237
in order to work with walnascar.

Except for refreshing the patch to apply, additional changes were needed:
* bitbake-worker filtered out exporting BB_MAKEFIFO to the environment, this
  may need a more elegant change in bb.data.exported_vars()
* make --jobserver-fds is replaced by jobserver-auth
* make not longer gives priority to -j in MAKEFLAGS over the command line
* This version autodetects make 4.4 and then uses --jobserver-auth=fifo:PATH
to pass a named pipe (preferred).

Make jobserver documented here: https://www.gnu.org/software/make/manual/html_node/POSIX-Jobserver.html

From the original patch description:

|This is a WIP to add a make job server into bitbake. This means the pool
|of make tokens is central and shared by all tasks executed instead of the
|current one job pool per task. Currently we can end up with many more
|make subprocesses executing that is intended or optimal.
|
|Implementation wise, make usually uses a pipe for this functionality. Here
|we substitute a named pipe (fifo) and intercept the make commands, passing
|in file descriptors to the central fifo. This assumes knowledge of make's
|internal API, on the plus side it hasn't changed since 1999.
|
|TODO:
|  * Remove hardcoded /tmp/makefifo and use something in TMPDIR or similar
|    (alongside the lock file?)
|  * Remove hardcoded make threads number and set from PARALLEL_MAKE
|  * If PARALELL_MAKE = "", don't set MAKEARGS (currently parallelism is
|    set everywhere) (need to check for -j in make commandline)

Signed-off-by: Trevor Gamblin <trevor.gamblin@windriver.com>
Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>

jobserver create queue in TMPDIR

Certains hosts may be running multiple bitbake's in parallel. Each
should have their own job queue.

Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>
---
 bin/bitbake-worker | 17 +++++++++--------
 lib/bb/runqueue.py | 26 +++++++++++++++++++++++++-
 2 files changed, 34 insertions(+), 9 deletions(-)

diff --git a/bin/bitbake-worker b/bin/bitbake-worker
index aa14ef191..15ed5c7c6 100755
--- a/bin/bitbake-worker
+++ b/bin/bitbake-worker
@@ -52,7 +52,7 @@ if sys.argv[1].startswith("decafbadbad"):
 try:
     if sys.stdout.name == '<stdout>':
         fl = fcntl.fcntl(sys.stdout.fileno(), fcntl.F_GETFL)
-        fl |= os.O_SYNC 
+        fl |= os.O_SYNC
         fcntl.fcntl(sys.stdout.fileno(), fcntl.F_SETFL, fl)
         #sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 0)
 except:
@@ -97,7 +97,7 @@ def worker_fire_prepickled(event):
 
 #
 # We can end up with write contention with the cooker, it can be trying to send commands
-# and we can be trying to send event data back. Therefore use a separate thread for writing 
+# and we can be trying to send event data back. Therefore use a separate thread for writing
 # back data to cooker.
 #
 worker_thread_exit = False
@@ -266,6 +266,7 @@ def fork_off_task(cfg, data, databuilder, workerdata, extraconfigdata, runtask):
                 the_data.setVar("BUILDNAME", workerdata["buildname"])
                 the_data.setVar("DATE", workerdata["date"])
                 the_data.setVar("TIME", workerdata["time"])
+                the_data.setVar("BB_MAKEFIFO", workerdata["makefifo"])
                 for varname, value in extraconfigdata.items():
                     the_data.setVar(varname, value)
 
@@ -288,14 +289,14 @@ def fork_off_task(cfg, data, databuilder, workerdata, extraconfigdata, runtask):
                     else:
                         logger.debug("Skipping disable network for %s since %s is not a local uid." % (taskname, uid))
 
-                # exported_vars() returns a generator which *cannot* be passed to os.environ.update() 
-                # successfully. We also need to unset anything from the environment which shouldn't be there 
+                # exported_vars() returns a generator which *cannot* be passed to os.environ.update()
+                # successfully. We also need to unset anything from the environment which shouldn't be there
                 exports = bb.data.exported_vars(the_data)
 
                 bb.utils.empty_environment()
                 for e, v in exports:
                     os.environ[e] = v
-
+                os.environ["BB_MAKEFIFO"] = workerdata["makefifo"]
                 for e in fakeenv:
                     os.environ[e] = fakeenv[e]
                     the_data.setVar(e, fakeenv[e])
@@ -337,7 +338,7 @@ def fork_off_task(cfg, data, databuilder, workerdata, extraconfigdata, runtask):
         else:
             profname = "profile-%s.log" % (fn.replace("/", "-") + "-" + taskname)
             prof = profile.Profile()
-            try: 
+            try:
                 ret = profile.Profile.runcall(prof, child)
             finally:
                 prof.dump_stats(profname)
@@ -401,7 +402,7 @@ class BitbakeWorker(object):
         self.extraconfigdata = None
         self.build_pids = {}
         self.build_pipes = {}
-    
+
         signal.signal(signal.SIGTERM, self.sigterm_exception)
         # Let SIGHUP exit as SIGTERM
         signal.signal(signal.SIGHUP, self.sigterm_exception)
@@ -419,7 +420,7 @@ class BitbakeWorker(object):
         signal.signal(signal.SIGTERM, signal.SIG_DFL)
         os.kill(os.getpid(), signal.SIGTERM)
 
-    def serve(self):        
+    def serve(self):
         while True:
             (ready, _, _) = select.select([self.input] + [i.input for i in self.build_pipes.values()], [] , [], 1)
             if self.input in ready:
diff --git a/lib/bb/runqueue.py b/lib/bb/runqueue.py
index a4689e134..96bc759dc 100644
--- a/lib/bb/runqueue.py
+++ b/lib/bb/runqueue.py
@@ -1401,6 +1401,7 @@ class RunQueue:
             "date" : self.cfgData.getVar("DATE"),
             "time" : self.cfgData.getVar("TIME"),
             "hashservaddr" : self.cooker.hashservaddr,
+            "makefifo" : self.cfgData.getVar("BB_MAKEFIFO"),
             "umask" : self.cfgData.getVar("BB_DEFAULT_UMASK"),
         }
 
@@ -1552,6 +1553,27 @@ class RunQueue:
 
         return bb.utils.better_eval(call, locs)
 
+    def setup_make_fifo(self):
+        fifoname = self.cfgData.getVar("TMPDIR") + "/makefifo"
+        self.cfgData.setVar("BB_MAKEFIFO", fifoname)
+        m = re.search(r'-j (\d+)', self.cfgData.getVar("PARALLEL_MAKE"))
+        if m:
+            threads = int(m.group(1))
+        else:
+            threads = 1
+
+        if os.path.exists(fifoname):
+            os.remove(fifoname)
+        os.mkfifo(fifoname)
+
+        # Has to be open for read and writing
+        self.makereadfd = os.open(fifoname, os.O_RDONLY|os.O_NONBLOCK)
+        self.makewritefd = os.open(fifoname, os.O_WRONLY)
+        wfd = os.fdopen(self.makewritefd, 'w')
+
+        for x in range(0, threads):
+            wfd.write('+')
+
     def _execute_runqueue(self):
         """
         Run the tasks in a queue prepared by rqdata.prepare()
@@ -1587,6 +1609,8 @@ class RunQueue:
             self.rqdata.init_progress_reporter.next_stage()
             bb.event.fire(bb.event.DepTreeGenerated(depgraph), self.cooker.data)
 
+            self.setup_make_fifo()
+
             if not self.dm_event_handler_registered:
                  res = bb.event.register(self.dm_event_handler_name,
                                          lambda x, y: self.dm.check(self) if self.state in [RunQueueState.RUNNING, RunQueueState.CLEAN_UP] else False,
@@ -1927,7 +1951,7 @@ class RunQueueExecute:
             self.max_loadfactor = float(self.max_loadfactor)
             if self.max_loadfactor <= 0:
                 bb.fatal("Invalid BB_LOADFACTOR_MAX %s, needs to be greater than zero." % (self.max_loadfactor))
-            
+
         # List of setscene tasks which we've covered
         self.scenequeue_covered = set()
         # List of tasks which are covered (including setscene ones)
-- 
2.43.0

